Unify waitpid handling

jfhg opened the pull request.

Araq commented on lib/pure/osproc.nim:
```diff
@@ -1014,28 +1013,30 @@ elif not defined(useNimRtl):
     const
       hasThreadSupport = compileOption("threads") and not defined(nimscript)
 
+    var status : cint = 1
```
A global variable? That's not re-entrant nor threadsafe.

jfhg replied:
Urgh, should be fixed now.

Araq commented:
What problems does this solve?

jfhg commented:
This PR tries to generally update p.exitCode only upon successful completion of waitpid().
In the current code this field is sometimes modified and then restored. Also in one case, the restore is only available under a when(freebsd) where as in other cases this is done generally.
This PR does not fix any bug I can reproduce. I came across the situation while debugging  #4792.

jfhg commented:
Sorry, it took me a while... I now added a test case that fails in current devel (tested on Linux and DragonFly BSD). To fix it I also added some missing exitStatus to exitCode conversions.
